packages/yarnpkg-pnpify/sources/Hoister.ts

```diff
@@ -0,0 +1,289 @@
+    const packages: PackageInfo[] = [];
+    const locators: PackageLocator[] = [];
+    const locatorToPackageMap = new Map<LocatorKey, PackageId>();
+    const packageInfos: PackageInformation<NativePath>[] = [];
+
+    const pnpRoots = pnp.getDependencyTreeRoots ? pnp.getDependencyTreeRoots() : [pnp.topLevel];
+
+    let lastPkgId = 0;
+
+    const getLocatorKey = (locator: PackageLocator): LocatorKey => `${locator.name}:${locator.reference}`;
+
+    const assignPackageId = (locator: PackageLocator, pkg: PackageInformation<NativePath>) => {
+      const locatorKey = getLocatorKey(locator);
+      let pkgId = locatorToPackageMap.get(locatorKey);
+      if (!pkgId) {
+        pkgId = lastPkgId;
+        lastPkgId++;
+        const packagePath = npath.toPortablePath(pkg.packageLocation);
+        let weight = 1;
+        if (this.options.optimizeSizeOnDisk && packagePath.indexOf(`.zip${NODE_MODULES_SUFFIX}`) > 0)
+          weight = fs.statSync(packagePath.split(NODE_MODULES_SUFFIX)[0]).size;
+
+        locatorToPackageMap.set(locatorKey, pkgId);
+        locators.push(locator);
+        packages.push({name: locator.name!, weight});
+        packageInfos.push(pkg);
+      }
+      return pkgId;
+    };
+
+    const getPackageId = (locator: PackageLocator) => locatorToPackageMap.get(getLocatorKey(locator));
+
+    const addedIds = new Set<PackageId>();
+    const addPackageToTree = (pkg: PackageInformation<NativePath>, pkgId: PackageId, parentDepIds: Set<PackageId>) => {
+      addedIds.add(pkgId);
+      const deps = new Set<PackageId>();
+      const peerDeps = new Set<PackageId>();
+      packageTree[pkgId] = {deps, peerDeps};
+
+      for (const [name, reference] of pkg.packageDependencies) {
+        if (reference) {
+          const locator = typeof reference === 'string' ? {name, reference} : {name: reference[0], reference: reference[1]};
+          const depPkgId = getPackageId(locator);
+          if (depPkgId) {
+            if (pkg.packagePeers) {
+              if (pkg.packagePeers.has(name)) {
+                peerDeps.add(depPkgId);
+              } else {
+                deps.add(depPkgId);
+              }
+            } else {
+              if (parentDepIds.has(depPkgId) && depPkgId !== pkgId) {
+                peerDeps.add(depPkgId);
+              } else {
+                deps.add(depPkgId);
+              }
+            }
+          }
+        }
+      }
+
+      const allDepIds = new Set(Array.from(deps).concat(Array.from(peerDeps)));
+      for (const depId of allDepIds) {
+        const depPkg = packageInfos[depId];
+        if (depPkg && !addedIds.has(depId)) {
+          addPackageToTree(depPkg, depId, allDepIds);
+        }
+      }
+    };
+
+    const assignedIds = new Set<PackageId>();
+    const assignPackageIds = (pkgId: PackageId) => {
+      assignedIds.add(pkgId);
+      const pkg = packageInfos[pkgId];
+      for (const [name, reference] of pkg.packageDependencies) {
+        if (reference) {
```

Comment:
Note: better explicitly check against `null` to avoid potential problems with empty strings (I'm considering adding a lint rule to guard against non-boolean non-object checks).